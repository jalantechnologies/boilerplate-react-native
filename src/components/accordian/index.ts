export { default as Accordion } from './accordian';
export { default as AccordionItem } from './accordianItem';
export type { AccordionProps, AccordionItemProps } from './types';
