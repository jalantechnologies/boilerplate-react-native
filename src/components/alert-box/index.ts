export { AlertProvider, useAlert } from './AlertProvider';
export { AlertType } from '../../types/alert';
