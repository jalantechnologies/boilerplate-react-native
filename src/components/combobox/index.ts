export { ComboboxInput } from './comboboxInput';
export { ComboboxOptionList } from './comboboxOptionList';
export { ComboboxOptionItem } from './comboboxOptionItem';
export * from './hooks/useCombobox';
export * from './types';
export * from './styles';
