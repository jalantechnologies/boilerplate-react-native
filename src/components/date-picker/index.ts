export { DatePicker } from './DatePicker';
export * from '../../types/date-picker';
