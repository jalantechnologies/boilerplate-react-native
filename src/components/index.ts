export { default as Brand } from './brand/brand';
export { default as ErrorFallback } from './error-fallback/error-fallback';
export { default as OTPInput } from './otp-input/otp-input';
