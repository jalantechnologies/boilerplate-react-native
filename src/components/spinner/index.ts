export { default as Spinner } from './spinner';
export type { SpinnerProps, SpinnerAnimationType } from './types';
export { DEFAULT_SPINNER_SIZE, DEFAULT_SPINNER_COLOR, DEFAULT_SPINNER_DURATION } from './constants';
