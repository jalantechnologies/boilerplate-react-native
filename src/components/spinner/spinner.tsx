import React, { useEffect, useRef } from 'react';
import { Animated, Easing } from 'react-native';

import { DEFAULT_SPINNER_SIZE, DEFAULT_SPINNER_COLOR, DEFAULT_SPINNER_DURATION } from './constants';
import type { SpinnerProps } from './types';

/**
 * Spinner Component - A loading indicator.
 *
 * Example usage:
 * ```tsx
 * <Spinner
 *   size={40}
 *   color="blue"
 *   animationType="indeterminate"
 *   duration={1000}
 * />
 * ```
 *
 * Note: Styling will be enhanced with NativeWind in the future.
 */
const Spinner: React.FC<SpinnerProps> = ({
  size = DEFAULT_SPINNER_SIZE,
  color = DEFAULT_SPINNER_COLOR,
  animationType = 'indeterminate',
  duration = DEFAULT_SPINNER_DURATION,
  style,
}) => {
  const rotateAnim = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    if (animationType === 'indeterminate') {
      Animated.loop(
        Animated.timing(rotateAnim, {
          toValue: 1,
          duration,
          easing: Easing.linear,
          useNativeDriver: true,
        }),
      ).start();
    } else if (animationType === 'fixed') {
      Animated.timing(rotateAnim, {
        toValue: 1,
        duration,
        easing: Easing.linear,
        useNativeDriver: true,
      }).start();
    }
  }, [animationType, duration, rotateAnim]);

  const spin = rotateAnim.interpolate({
    inputRange: [0, 1],
    outputRange: ['0deg', '360deg'],
  });

  // Basic placeholder styling for now.
  const spinnerStyle = {
    width: size,
    height: size,
    borderWidth: size / 10,
    borderColor: color,
    borderRadius: size / 2,
    borderTopColor: 'transparent', // creates a spinner effect
    transform: [{ rotate: spin }],
  };

  return <Animated.View style={[spinnerStyle, style]} />;
};

export default Spinner;
