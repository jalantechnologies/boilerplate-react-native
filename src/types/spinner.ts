export enum SpinnerTypes {
  PRIMARY = 'primary',
  SECONDARY = 'secondary.50',
  TERTIARY = 'tertiary',
}

export enum SpinnerSize {
  SMALL = 'small',
  LARGE = 'large',
}
