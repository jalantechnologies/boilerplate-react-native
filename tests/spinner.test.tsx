import { render } from '@testing-library/react-native';
import React from 'react';

import { Spinner } from '../src/components/spinner';

jest.useFakeTimers();

describe('Spinner Component', () => {
  test('renders correctly with default props', () => {
    const { toJSON } = render(<Spinner />);
    expect(toJSON()).toMatchSnapshot();
  });

  test('applies custom size and color', () => {
    const customSize = 60;
    const customColor = 'red';
    const { toJSON } = render(<Spinner size={customSize} color={customColor} />);
    const tree = toJSON();
    // Check that width, height, and borderColor are applied
    if (!tree || Array.isArray(tree) || !('props' in tree)) {
      throw new Error('Unexpected tree structure');
    }
    const styleArray = Array.isArray(tree.props.style) ? tree.props.style : [tree.props.style];
    expect(styleArray).toEqual(
      expect.arrayContaining([
        expect.objectContaining({
          width: customSize,
          height: customSize,
          borderColor: customColor,
        }),
      ]),
    );
  });

  test('runs indeterminate animation mode', () => {
    const { toJSON } = render(<Spinner animationType="indeterminate" />);
    const tree = toJSON();
    if (!tree || Array.isArray(tree) || !('props' in tree)) {
      throw new Error('Unexpected tree structure');
    }
    const styles = Array.isArray(tree.props.style) ? tree.props.style : [tree.props.style];
    const hasRotate = styles.some((style: any) => style.transform && style.transform[0].rotate);
    expect(hasRotate).toBe(true);
  });

  test('runs fixed animation mode', () => {
    const { toJSON } = render(<Spinner animationType="fixed" duration={2000} />);
    const tree = toJSON();
    if (!tree || Array.isArray(tree) || !('props' in tree)) {
      throw new Error('Unexpected tree structure');
    }
    const styles = Array.isArray(tree.props.style) ? tree.props.style : [tree.props.style];
    const hasRotate = styles.some((style: any) => style.transform && style.transform[0].rotate);
    expect(hasRotate).toBe(true);
  });
});
