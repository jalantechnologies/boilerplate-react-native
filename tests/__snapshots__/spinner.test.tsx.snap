// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner Component renders correctly with default props 1`] = `
<View
  collapsable={false}
  style={
    {
      "borderColor": "#000",
      "borderRadius": 20,
      "borderTopColor": "transparent",
      "borderWidth": 4,
      "height": 40,
      "transform": [
        {
          "rotate": "0deg",
        },
      ],
      "width": 40,
    }
  }
/>
`;
